feat(reaper): Add process reaper support

stevenh opened the pull request:
##### Summary
Add a child process reaper to the main netdata app if running as init (pid = 1).

This prevents zombie processes when a child is re-parented to netdata when its running in a container.

Also:
* Few style cleanups to match surrounding code.

Fixes: #6033

##### Component Name
netdata binary

##### Additional Information
This re-purposes old commented out code in `popen.c`, which already implemented part of the required process tracking.

Without this on a standard netdata docker install we saw at least one zombie `timeout` process straight after the container was started.

mfundul commented on libnetdata/popen/popen.c:
```diff
@@ -128,6 +149,60 @@ static inline FILE *custom_popene(const char *command, volatile pid_t *pidptr, c
+    netdata_mutex_lock(&myp_lock);
+    for (mp = mypopen_root; mp != NULL; mp = next) {
+        next = mp->next;
+        free(mp);
```
since we will change `malloc()` to `mallocz()` this should become `freez()`

mfundul commented on daemon/signals.c:
```diff
@@ -157,6 +229,11 @@ void signals_handle(void) {
 
+                            case NETDATA_SIGNAL_CHILD:
+                                if (reaper_enabled == 1)
+                                    reap_child();
```
I think that at this point, there may have been multiple `SIGCHLD` signals before `signals_waiting[i].count = 0;` was set.

I think a single `reap_child()` call is not enough and that we need a loop. What do you think?

stevenh replied:
Yes indeed, didn't spot this wasn't the actual signal handler.

I would prefer to actual rework the signal handlers to be inline, otherwise you've going to have to deal with races. This would should also be more performant thoughts?

mfundul replied:
I'm afraid that is not going to work because in the signal handlers themselves you can do almost nothing, and netdata already does **a lot** of heavy lifting in `signals_handle()`. You cannot even use mutexes such as those from `myp_reap()` in `reap_child()`. See [here](http://man7.org/linux/man-pages/man7/signal-safety.7.html) for reference.

`signals_handle()` is being called by the netdata daemon master thread, that is possibly the one with `pid == 1` in the container case, and it's devoted to exclusively handle signals after the initialization. Due to the limitations of signal safety the scope of a perfect solution would be too big in relation to the motivation behind this pull request.

I would prefer you figure out  a way to catch the signals without making a lot of changes.

stevenh replied:
Hmm actually looks like this was a conscious decision, to avoid long running signal handlers. I will take a different approach.

stevenh replied:
Yep already spotted that when investigating the options, but thanks for the heads up.

mfundul commented on libnetdata/popen/popen.c:
```diff
@@ -2,46 +2,66 @@
+                mp->prev->next = mp->next;
+            if (mypopen_root == mp)
+                mypopen_root = mp->next;
+            free(mp);
```
since we will change `malloc()` to `mallocz()` this should become `freez()`

mfundul commented on libnetdata/popen/popen.c:
```diff
@@ -2,46 +2,66 @@
+        }
     }
+
+    if (mp != NULL)
```
Shouldn't that be `if (mp == NULL)` ?

ilyam8 commented:
@stevenh i would suggest to not force-push w/o reason, it is not clear what is fixed and we have to review from scratch.

mfundul commented on libnetdata/popen/popen.c:
```diff
@@ -2,46 +2,66 @@
+        return;
+
+    mp = mallocz(sizeof(struct mypopen));
+    if (mp == NULL) {
```
Forgot to mention, `mallocz()` already does the `fatal()` inside so you can remove this `if` statement.

stevenh commented:
Thanks for the feedback so far guys, always good to get to know the nuances of project, no force pushes. 👍 

I've addressed all the feedback, fixing a few other niggles along the way, as well as eliminating some robustness issues which became apparent when stress testing the for multiple children exiting; i.e. the data race when setting pid.

mfundul commented on daemon/signals.c:
```diff
@@ -100,6 +122,76 @@ void signals_reset(void) {
+        } else if (myp_reap(i.si_pid) == 0) {
+            // myp managed, sleep for a short time to avoid busy wait while
+            // this is handled by myp.
+            usleep(10000);
```
I really hate this but I don't see a way around it without rewriting the daemon.

mfundul replied:
A potential problem here is that all signal handling is effectively blocked until somebody calls `mypclose()`, such as health reload or exiting.

stevenh replied:
Indeed wasn't a fan either, the only thing I thought of was to have this handled in a separate thread; seems like this could be a future enhancement?

Shame its not in `go` as that would just be a few lines of code, not so simple in `C` though ;-)

mfundul replied:
Even in `go` I don't think it's guaranteed that `waitid(P_ALL, (id_t)0, &i, WEXITED|WNOHANG|WNOWAIT)` will return any other pid than the one being handled by `myp` which would not allow the loop to iterate.

stevenh replied:
I actually meant its much easier to throw up a `goroutine` to handle each type of signal, not that it solves the wait issue.

mfundul commented on daemon/signals.c:
```diff
@@ -100,6 +122,76 @@ void signals_reset(void) {
+        break;
+    default:
+        error("SIGNAL: Child %d gave us a SIGCHLD with code %d and status %d.", pid, i.si_code, i.si_status);
+    }
```
I'm not sure we want to print to the netdata log all reaped processes of a container that happen to be re-parented to the netdata process. I feel like this should be a silent operation.

mfundul commented on daemon/signals.c:
```diff
@@ -100,6 +122,76 @@ void signals_reset(void) {
+    siginfo_t i;
+
+    errno = 0;
+    info("SIGNAL: Reaping pid: %d...", pid);
```
Same as below, I think we should remove the prints.

stevenh replied:
How about switching to debug, as they may be useful if there's a bad plugin or something?

mfundul replied:
Sure, you can switch to debug.

cakrit commented:
We appreciate the work @reaper!!!

ilyam8 commented:
@stevenh the reaper 😄
